### PR TITLE
property set data and resource providers

### PR DESCRIPTION
**Pull request by rajagopalans**
No tests yet.

### CODE REVIEW

**Review by chrismarget-j — APPROVED** [review]

File: apstra/resource_property_set.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	var api *apstra.PropertySet
+	var ace apstra.ApstraClientErr
+
+	api, err = o.client.GetPropertySet(ctx, apstra.ObjectId(state.Id.ValueString()))
+	if err != nil && errors.As(err, &ace) && ace.Type() == apstra.ErrNotfound {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("id"),
+			"PropertySet not found",
+			fmt.Sprintf("PropertySet with ID %q not found", state.Id.ValueString()))
+		return
```
Comment (chrismarget-j) [inline apstra/resource_property_set.go:87]: This should just invoke `resp.State.RemoveResource()`, no need for an error.

File: apstra/apstra_validator/parse_json.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+type ParseJsonValidator struct{}
+
+type ParseJsonValidatorRequest struct {
+	Config      tfsdk.Config
+	ConfigValue attr.Value
+}
+
+type ParseJsonValidatorResponse struct {
+	Diagnostics diag.Diagnostics
+}
```
Comment (chrismarget-j) [inline apstra/apstra_validator/parse_json.go:24]: `ParseJsonValidatorRequest` and `ParseJsonValidatorResponse` aren't used, can be deleted.

File: apstra/design/property_set.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			Validators: []validator.String{
+				stringvalidator.LengthAtLeast(1),
+				stringvalidator.ExactlyOneOf(path.Expressions{
+					path.MatchRelative(),
+					path.MatchRoot("name"),
+				}...),
+			},
+		},
+		"name": dataSourceSchema.StringAttribute{
+			MarkdownDescription: "Populate this field to look up a property set by name. Required when `id` is omitted.",
```
Comment (chrismarget-j) [inline apstra/design/property_set.go:40]: Please make these comments consistent with capitalization. Apstra objects in these doc strings are capitalized: "Property Set"

Several instances in this file.

File: apstra/design/property_set.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+func (o *PropertySet) LoadApiData(ctx context.Context, in *apstra.PropertySetData, diags *diag.Diagnostics) {
+	o.Label = types.StringValue(in.Label)
+	var d diag.Diagnostics
+	o.Blueprints, d = types.SetValueFrom(ctx, types.StringType, in.Blueprints)
+	diags.Append(d...)
+	o.Values = types.StringValue(string(in.Values))
+}
+
+func (o *PropertySet) Request(ctx context.Context, diags *diag.Diagnostics) *apstra.PropertySetData {
```
Comment (chrismarget-j) [inline apstra/design/property_set.go:92]: `ctx` and `diags` aren't used. I'd like to keep them for consistency with the other `Request()` methods, but they can be reduced to `_` when unused.

File: apstra/utils/compare_jsons.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		if err != nil {
+			d.AddError("error unmarshalling string", err.Error())
+			return false
+		}
+	}
+	return reflect.DeepEqual(map1, map2)
+}
+
+// IsJSON takes a string and returns true if json, false if not
+
```
Comment (chrismarget-j) [inline apstra/utils/compare_jsons.go:38]: Extraneous blank line breaks GoDoc notation.

**Review by chrismarget-j — CHANGES_REQUESTED** [review]
inline comments scattered around.

Also, need docs generated before the release process can run.

File: apstra/resource_property_set.go (hunk 1 above)
Comment (rajagopalans) [reply apstra/resource_property_set.go:87]: done

File: apstra/apstra_validator/parse_json.go (hunk 2 above)
Comment (rajagopalans) [reply apstra/apstra_validator/parse_json.go:24]: done

File: apstra/design/property_set.go (hunk 3 above)
Comment (rajagopalans) [reply apstra/design/property_set.go:40]: Done

File: apstra/design/property_set.go (hunk 4 above)
Comment (rajagopalans) [reply apstra/design/property_set.go:92]: Done

Comment (chrismarget-j) [reply apstra/design/property_set.go]: missing commit? They're still named `ctx` and `diags` here.

**Comment by rajagopalans** [comment]
No idea why. No pending commits and local copy has _

File: apstra/design/property_set.go (hunk 4 above)
Comment (chrismarget-j) [reply apstra/design/property_set.go]: looks okay to me now. Git pull just collected:
```
   7330bc7..0dc6c0e  propertysets -> origin/propertysets
```

Weird. I don't know why I saw this in the web view.

File: apstra/utils/compare_jsons.go (hunk 5 above)
Comment (rajagopalans) [reply apstra/utils/compare_jsons.go]: fixed

**Review by chrismarget-j — APPROVED** [review]

**Pull request merged**